### PR TITLE
Changed is_connected to return bool while maintaining error checking

### LINKED ISSUE #109: is_connected raises MMQTTException when not connected, rather that returning False.
I would expect is_connected to return True or False, rather than raise an exception in the False case:
```
>>> mqtt_client.is_connected()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "adafruit_minimqtt/adafruit_minimqtt.py", line 978, in is_connected
MMQTTException: MiniMQTT is not connected.
```

To recreate:
```
import socketpool
import wifi
import adafruit_minimqtt.adafruit_minimqtt as MQTT
from secrets import secrets
pool = socketpool.SocketPool(wifi.radio)
mqtt_client = MQTT.MQTT(
    broker=secrets["broker"],
    port=secrets["port"],
    socket_pool=pool,
)
mqtt_client.is_connected()

```

### PR DESCRIPTION
**Pull request by BiffoBear**
closes #109 
Renamed original is_connected method to _connected to maintain internal connection status checking. Created a new is_connected function that returns a bool. This is a breaking change.

This change allows code to be written like:
```
while not mqtt_client.is_connected():
    try:
        mqtt_client.connect()
    except RuntimeError:
        print("Unable to connect. Waiting for 10 seconds before retrying…")
        time.sleep(10)
print("Connected to MQTT broker.")
```

### CODE REVIEW

**Review by tekktrik — COMMENTED** [review]
I think this is a good change, but I'm also not as familiar with MQTT applications.  I think this is worth getting a second opinion since its also API breaking.  I'll bring it up at the CircuitPython weekly meeting tomorrow!

**Comment by BiffoBear** [comment]
> I think this is a good change, but I'm also not as familiar with MQTT applications. I think this is worth getting a second opinion since its also API breaking. I'll bring it up at the CircuitPython weekly meeting tomorrow!

After poking around in the code, I believe the `is_connected()` method is poorly named. The method does not reflect the state of the connection when it is called. It checks whether a network socket exists and whether an initial connection was successfully made to the MQTT broker when `MQTT.connect()` was called. The `is_connected` method acts as a gatekeeper so methods like `publish()` and `ping()` are not called before the connection to the broker has been made. This is why it returns `True` or raises an exception.

If the MQTT broker is down, the `is_connected` method will return `True` but any attempt to write to the broker will raise a `BrokenPipeError: 32` exception.

After poking the code a bit, `MQTT.connect()` returns a value if a connection with the broker was successfully initiated and raises exceptions for failed connections. Thus a `while not mqtt_client.is_connected:` loop is not required.
In short, I no longer think that this breaking change is worth making.

Perhaps the method should be renamed `_initial_connection_made()` or something along those lines to make clear that it does not reflect the current status and is not particularly useful to the user.

**Review by jepler — CHANGES_REQUESTED** [review]
Thanks!  However, I think it could be written more simply. Do you mind changing it around?

File: adafruit_minimqtt/adafruit_minimqtt.py [hunk 1]
```diff
@@ -1046,14 +1046,23 @@ def _valid_qos(qos_level):
         else:
             raise MMQTTException("QoS must be an integer.")
 
-    def is_connected(self):
+    def _connected(self):
```
[the remark this answers is not in the document]
Comment (BiffoBear) [reply adafruit_minimqtt/adafruit_minimqtt.py:1049]: Thanks, I see what you mean and your code is more elegant. I don't think we should change the behaviour though.

When I read the issue thread I had barely used the library, so I assumed that `is_connected` checked the current state of the connection to the MQTT broker. It does not. It checks to see whether `connect()` has been called before deciding whether code should be run and raises an exception if it should not. e. g. if we try to `publish()` before connecting.

The function behaves as designed. The problem is with the function's name.

`connect()` either connects successfully or raises an exception, so no need for the user to call `is_connected` in a while loop (which was what I thought I should do).

If the MQTT broker is down, the `is_connected` method will return `True` but any attempt to write to the broker will raise a `BrokenPipeError: 32` exception.

I propose that I close this pull request without merging any changes.

I could open a new pull request that renames the `is_connected` function to something like `_initial_connection_made` for a breaking change or one that edits the doc-string to make the function's purpose clear.

[the remark this answers is not in the document]
Comment (jepler) [reply adafruit_minimqtt/adafruit_minimqtt.py:1049]: Maybe I misunderstood your change, because I thought yours and mine would function identically.

 * is_connected returns a bool, like yours
 * _connected throws an exception if not is_connected, like yours

can you help me understand where you're seeing it as different? The return value of my _connected is None, not True, but as the return value is never used it didn't seem to matter as a difference.

[the remark this answers is not in the document]
Comment (BiffoBear) [reply adafruit_minimqtt/adafruit_minimqtt.py:1049]: Hi, your change and mine do function the same. I'm not sure we should make the change at all. 

Here's why:

The `is_connected()` method only checks to see if there is a socket and if the variable `self._is_connected' is true.

`self._is_connected` is initialised `False` then set to `True` when `connect()` is called and a connection to the broker is successful. `self._is_connected` is set to `False` when `disconnect() is called.

The code calls 'is_connected()' at the beginning of methods like 'disconnect()`, 'publish()' and `will_set()`.

In `will_set`, for example:

```python
    if self._is_connected:
        raise MMQTTException("Last Will should only be called before connect().")
```

From reading through the code, I conclude that the `is_connected()` method has nothing to say about whether there is connection to the MQTT broker at the moment when it is called. The method is used by the author to decide whether it is appropriate to call methods based on whether a connection to whether `connect()` or `disconnect()` have previously been called.

Because the method is named `is_connected` I and others incorrectly assumed that the method will tell us whether there is a current, live connection to the MQTT server and think that the method should return `True` instead of raising an exception. We're wrong.

Therefore I think that changing `is_connected()` to return `True` instead of raising an exception will cause more problems for users than leaving it alone. The `ping()` method is the intended way to see if the broker connection is live or not.

I propose to cancel this pull request and open a new one to have a conversation around whether the `is_connected()` should be renamed to indicate that it is a private function, not intended to be called by from outside the module.

**Comment by BiffoBear** [comment]
Thanks to @2bndy5 for clarifying my thinking and pointing out that the Paho implementation returns a `bool`. Changes requested made.

File: adafruit_minimqtt/adafruit_minimqtt.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (BiffoBear) [reply adafruit_minimqtt/adafruit_minimqtt.py:1049]: Requested changes made, ready to go.

**Review by brentru — APPROVED** [review]
